### DISCUSSION TITLE
Metamodel Domain Boundaries

### DISCUSSION THREAD

**Discussion by lwriemen** [modelint/shlaer-mellor-metamodel#61] [General]
I saw that Subsystem was added to the shlaer-mellor-metamodel, and this makes me question how you're defining the Shlaer-Mellor Metamodel domain. I would consign Subsystem to the Representation Domain (for lack of a better quick name). In graphical models, a subsystem is definitely a nice way to constrain the amount of information one must view at a time in the domain object model, but it isn't a necessity for conforming to the rules of creating a domain model. In my metamodel, I intentionally restricted the metamodel to the rules of creating a Shlaer-Mellor domain model.

While the literature to date talks about notations and recursive design, the focus of the method is domain modeling. I feel that the Shlaer-Mellor Metamodel should retain this focus.

**Comment by modelint** [comment]
Hi Lee,

This is an excellent point. I have grappled with the decision to include subsystem partitioning here on more than one occasion and am currently leaning toward keeping subsystems as an integral part of the metamodel. But I can certainly appreciate your perspective on this. You may ultimately bring me back to your point of view, but I'm not currently convinced.  Here's my take on it:

I think we both understand that subsystem partitions are purely for management purposes. In the functioning system, they make no difference. You can carve up a given domain into many different subsystem partitions and still have the same system in the end. On the other hand, choices of how you split out your domains, abstract your classes and so forth do have a significant impact on your system description and its functionality.

That said, I have to reflect on a number of projects where I have large domains with multiple subsystems. We can count the SM Metamodel in this category, but certainly my Toyota driving environment work and even Flatland benefit greatly from a coherent subsystem partitioning scheme. For management of large domains, subsystems are essential and I suspect we agree on this point as well.

I fear that if subsystems are not supported  as near to first class citizens in our modeling approach, subsystems will be either ignored or poorly defined. I have seen a lot of large domains thrown together as incomprehensible spider diagrams more than once. Subsystem partitioning, is of course, ripe for abuse like any other construct. But applied properly, I think it helps organize the thinking behind a domain in subtle, but important and underappreciated ways. Not as important as domain partitioning by any stretch.

The concept of a representational domain as you call it leaves open the possibility that there are many different ways to organize a domain and perhaps other groups of modeling constructs. Sally liked to use the term 'clusters' for smaller class/relationship groups. I'm not so sure there was ever much value in that particular kind of grouping. So I think this basis for this domain is a bit tenuous.

For lack of a better approach, I will keep subsystems as is, but am open to alternatives down the road.

– Leon

**Reply by lwriemen** [reply]
I intentionally avoided the term, "Graphics Domain", because textual representation is certainly a possibility. While "this basis", Subsystem, is certainly not the only basis, e.g., class, relationship, domain, state, etc., I will argue it belongs there, because it might make more sense in a graphical representation, than in a textual representation. Adding Subsystem to the metamodel introduces a partitioning that might require a lot of spurious rules.

I saw a lot of confusion with the term, "Subsystem", over the years in the various Shlaer-Mellor discussions. Another term it caused in BridgePoint, Imported Class, added to the confusion. Anyway, we can see how it works out for you as your metamodel progresses.
